src/main/java/com/google/googleidentity/oauth2/token/InMemoryOAuth2TokenService.java

```diff
@@ -0,0 +1,451 @@
+                            .setUsername(request.getRequestAuth().getUsername())
+                            .setIsScoped(request.getRequestBody().getIsScoped())
+                            .addAllScopes(request.getRequestBody().getScopesList())
+                            .build();
+
+            userClientRefreshTokenMap.get(username).put(clientID, refreshToken);
+
+            builder.setRefreshToken(refreshToken.getRefreshToken());
+        }
+
+        OAuth2AccessToken token = builder.build();
+
+        accessTokenMap.put(accessTokenValue, token);
+        return token;
+    }
+
+    public Optional<OAuth2AccessToken> refreshToken(String refreshToken) {
+        Optional<OAuth2RefreshToken> token = readRefreshToken(refreshToken);
+
+        if (!token.isPresent()) {
+            return Optional.empty();
+        }
+
+        String username = token.get().getUsername();
+        String clientID = token.get().getClientId();
+
+        HashMap<String, OAuth2AccessToken>  accessTokenMap=
+                    userClientAccessTokenMap.get(username).get(clientID);
+
+        String accessTokenValue = UUID.randomUUID().toString();
+
+        while (accessTokenMap.containsKey(accessTokenValue)) {
+            accessTokenValue = UUID.randomUUID().toString();
+        }
+
+        UserClientTokenInfo info =
+                new UserClientTokenInfo(username, clientID, accessTokenValue);
+
+        OAuth2AccessToken newToken =
+                OAuth2AccessToken.newBuilder()
+                        .setAccessToken(info.getTokenString())
+                        .setClientId(clientID)
+                        .setUsername(username)
+                        .setIsScoped(token.get().getIsScoped())
+                        .addAllScopes(token.get().getScopesList())
+                        .setExpiredTime((new Date().getTime())/1000l + tokenValidTime)
+                        .setRefreshToken(refreshToken)
+                        .build();
+
+        accessTokenMap.put(accessTokenValue, newToken);
+
+        return Optional.ofNullable(newToken);
+    }
+
+    public Optional<OAuth2AccessToken> readAccessToken(String accessToken) {
+        UserClientTokenInfo info = null;
+
+        try{
+            info = new UserClientTokenInfo(accessToken);
+        } catch (InvalidParameterException exception) {
+            log.log(Level.INFO,  "Invalid refresh token Value", exception);
+            return Optional.empty();
+        }
+
+        String username = info.getUsername();
+        String clientID = info.getClientID();
+        String tokenValue = info.getTokenValue();
+
+        if (userClientAccessTokenMap.containsKey(username) &&
+                userClientAccessTokenMap.get(username).containsKey(clientID)) {
+            return Optional.ofNullable(
+                    userClientAccessTokenMap.get(username).get(clientID).get(tokenValue));
+        } else {
+            return Optional.empty();
```

Comment:
Same here, this logic should be shared between readAccessToken and readRefreshToken.